## Data Loading & Cleaning Phase :

In [70]:
import pandas as pd
import os
data_dir = '../data/raw'
files = ['dataset-tickets-multi-lang-4-20k.csv', 'dataset-tickets-multi-lang3-4k.csv', 'aa_dataset-tickets-multi-lang-5-2-50-version.csv']
df_temp = [pd.read_csv(os.path.join(data_dir, f)) for f in files]
df = pd.concat(df_temp, ignore_index=True)

## We loaded the csv file and we will process it according to our needs ( by removing certain unecessary things basically cleaning it to need and then saving it )

In [72]:
print(df.shape)

(52587, 18)


In [71]:
print(df.columns.tolist())
df.isnull().sum()
df['language'].value_counts()   # if this column exists

['subject', 'body', 'answer', 'type', 'queue', 'priority', 'language', 'tag_1', 'tag_2', 'tag_3', 'tag_4', 'tag_5', 'tag_6', 'tag_7', 'tag_8', 'business_type', 'tag_9', 'version']


language
en    29652
de    21174
es      812
fr      476
pt      473
Name: count, dtype: int64

In [73]:
df = df[df['language'] == 'en']
print(df.shape)

(29652, 18)


In [ ]:
df['type'].value_counts()

priority
medium    12127
high      11512
low        6013
Name: count, dtype: int64

In [76]:
df['queue'].value_counts()


queue
Technical Support                  8617
Product Support                    5540
Customer Service                   4482
IT Support                         3500
Billing and Payments               3017
Returns and Exchanges              1467
Service Outages and Maintenance    1157
Sales and Pre-Sales                 885
Human Resources                     568
General Inquiry                     419
Name: count, dtype: int64

In [77]:
df['priority'].value_counts()

priority
medium    12127
high      11512
low        6013
Name: count, dtype: int64

In [78]:
df['type'].value_counts()

type
Incident    11803
Request      8537
Problem      6163
Change       3149
Name: count, dtype: int64

In [79]:
pd.set_option('display.max_colwidth', None)
for q, group in df.groupby('queue'):
    print(f"\n=== {q} ===")
    print(group['body'].sample(1, random_state=1).values[0])


=== Billing and Payments ===
Hello Customer Support, I am reaching out to inquire about the billing features of the SaaS project management tool. Could you provide information on the available plans, payment options, and any discounts that may be offered? I am particularly interested in learning about the scalability of the tool and how it can adapt to growing business needs. Additionally, I would appreciate details on the level of support provided and the security measures in place to protect user data. Thank you for your time, and I look forward to hearing back soon.

=== Customer Service ===
Customer Support, seeking information on integrating Smart-Tracker project management tools. Could provide details on the process? Would greatly appreciate any guidance or documentation you can provide. Specifically, I would like to know which tools are compatible with Smart-Tracker and the steps to integrate them. I believe that integrating Smart-Tracker tools would help streamline my project 

In [80]:
df.isnull().sum()

subject           3810
body                 2
answer               6
type                 0
queue                0
priority             0
language             0
tag_1                0
tag_2               16
tag_3              113
tag_4             2584
tag_5            12173
tag_6            20993
tag_7            25589
tag_8            27937
business_type    28261
tag_9            29652
version          13314
dtype: int64

In [81]:
df['queue'].value_counts()


queue
Technical Support                  8617
Product Support                    5540
Customer Service                   4482
IT Support                         3500
Billing and Payments               3017
Returns and Exchanges              1467
Service Outages and Maintenance    1157
Sales and Pre-Sales                 885
Human Resources                     568
General Inquiry                     419
Name: count, dtype: int64

In [82]:
df['body'].str.len().describe()


count    29650.000000
mean       386.158078
std        231.028312
min          4.000000
25%        209.000000
50%        372.000000
75%        539.000000
max       2466.000000
Name: body, dtype: float64

In [83]:
df.duplicated(subset='body').sum()

np.int64(4596)

In [84]:
pd.set_option('display.max_colwidth', None)
df[['subject','body','answer']].dropna().sample(3, random_state=1)

,subject,body,answer
15743,Strategies for Enhancing Tech Product Brand Growth Digitally,I am contacting you to seek guidance on efficient digital strategies that can aid in the growth of a tech product brand. Would you be able to share any recommendations or insights?,"Certainly, I can assist you with digital strategies to enhance the growth of your tech product brand. We could explore avenues such as social media marketing, content creation, and influencer collaborations. For personalized recommendations, could you provide details on your existing brand strategy and objectives? If convenient, we could arrange a call at <tel_num> to delve deeper into this."
18670,Inquiry on Zapier Integrations with Investment Analytics Tools,"I am contacting you to seek information regarding the integrations available between Zapier and various investment analytics tools. Could you kindly furnish me with a list of the tools that are currently supported? I am especially interested in knowing about the integrations that facilitate automated data synchronization and workflow optimization. Furthermore, it would be greatly helpful if you could share any details about the setup and configuration process for these integrations. Thank you for your time and assistance in this matter. I eagerly await your response.",I will send you a list of supported investment analytics tools via reply email.
10285,Procedures for Safeguarding Medical Information,I would like to know about the precautions your hospital takes to ensure the confidentiality of medical data within its IT infrastructure. Could you elaborate on the steps you have put in place to safeguard patient data?,"Patient data security is a top priority for us. Our hospital has put in place comprehensive security protocols, such as encryption and access controls, to ensure the protection of patient information. I will send you a detailed document that outlines our security procedures. If you wish to discuss this further, I can arrange a call at your convenience."


In [ ]:
df = df.drop(columns=['tag_1','tag_2','tag_3','tag_4','tag_5','tag_6','tag_7','tag_8','tag_9','business_type','version','language'])
# drop unnecessary columns

In [ ]:
df = df.dropna(subset=['subject','body','answer']) 
# drop missing things

In [88]:
df = df.drop_duplicates(subset='body')
# drop duplicate bodies

In [89]:
df['text_input'] = df['subject'] + " " + df['body']
#Combine subject + body for classifier input

In [1]:
from triage.data.cleaning import clean_pipeline, save_processed_data

df = clean_pipeline()
save_processed_data(df)
df.shape

(22053, 7)

In [93]:
print(df.shape)
df.isnull().sum()

(22053, 7)


subject       0
body          0
answer        0
type          0
queue         0
priority      0
text_input    0
dtype: int64

In [94]:
df.to_csv(r'C:\Users\jenil\OneDrive\Desktop\Machine learning\AI-Ops\triage\data\processed/tickets_clean.csv')

THINGS I PERFORMED IN THIS NOTEBOOK :


Explored raw dataset — loaded and merged 3 CSV files (~29,650 English tickets), checked shape, nulls, and class balance across type, queue, and priority.

Validated label quality — read actual ticket text against queue labels to confirm real signal exists (after an earlier dataset attempt turned out to have synthetic, disconnected labels).

Dropped unusable columns — removed sparse tag columns, business_type, version, and non-English rows.

Cleaned text data — removed rows with missing subject/body/answer, dropped ~4,596 duplicate ticket bodies, stripped HTML tags and placeholder tokens (e.g. <br>, <tel_num>).

Engineered classifier input & saved output — combined subject + body into text_input, confirmed answer column as a valid RAG knowledge source, saved final clean dataset (22,053 rows, zero nulls) to data/processed/tickets_clean.csv.